src/internal/document_api/mod.rs

```diff
@@ -435,53 +500,169 @@ pub fn encrypt_document<
     grant_to_author: bool,
     user_grants: &'a Vec<UserId>,
     group_grants: &'a Vec<GroupId>,
-    policy_grant: Option<&PolicyGrant>,
+    policy_grant: Option<&'a PolicyGrant>,
 ) -> impl Future<Item = DocumentEncryptResult, Error = IronOxideErr> + 'a {
     let (dek, doc_sym_key) = transform::generate_new_doc_key(recrypt);
     let doc_id = document_id.unwrap_or(DocumentId::goo_id(rng));
+    aes::encrypt_future(rng, &plaintext.to_vec(), *doc_sym_key.bytes())
+        .join(resolve_keys_for_grants(
+            auth,
+            user_grants,
+            group_grants,
+            policy_grant,
+            if grant_to_author {
+                Some(&user_master_pub_key)
+            } else {
+                None
+            },
+        ))
+        .and_then(move |(encrypted_doc, (grants, key_errs))| {
+            encrypt_document_core(auth, recrypt, dek, encrypted_doc, &doc_id, grants)
+                .into_future()
+                .and_then(move |r| {
+                    let encryption_errs = r.encryption_errs.clone();
+                    document_create(
+                        &auth,
+                        r,
+                        doc_id,
+                        &document_name,
+                        [key_errs, encryption_errs].concat(),
+                    )
+                })
+        })
+}
+
+type UserMasterPublicKey = PublicKey;
+/// Get the public keys for a document grant.
+///
+/// # Arguments
+/// `auth`          - info to make webservice requests
+/// `user_grants`   - list of user ids to which document access should be granted
+/// `group_grants`  - list of groups ids to which document access should be granted
+/// `policy_grant`  - policy to apply for document access
+/// `grant_to_author` - true if document access should be granted to the logged in user, else false
```

Comment:
Is this correct? The type doesn't seem to line up with the last argument in the function (`maybe_user_master_pub_key`)

Reply:
Thanks. Fixed.